crane/app_util.py

```diff
@@ -199,3 +218,85 @@ def validate_and_transform_repoid(repo_id):
     if repo_id.startswith('library/'):
         return repo_id[len('library/'):]
     return repo_id
+
+
+def name_is_authorized(name):
+    """
+    Determines if the current request is authorized to read the given repo name.
+
+    :param name: name of the repository being accessed
+    :type  name: basestring
+
+    :raises exceptions.HTTPError: if authorization fails
+                                  403: if the user is not authorized
+                                  404: if the repo does not exist in this app
+    """
+    v2_response_data = get_v2_data()
+    v2_repo_tuple = v2_response_data['repos'].get(name)
+
+    # if this deployment of this app does not know about the requested repo
+    if v2_repo_tuple is None:
+        raise exceptions.HTTPError(httplib.NOT_FOUND)
+
+    if v2_repo_tuple.protected:
+        cert = _get_certificate()
+        if not cert or not cert.check_path(v2_repo_tuple.url_path):
+            # return 404 so we don't reveal the existence of repos that the user
+            # is not authorized for
+            raise exceptions.HTTPError(httplib.NOT_FOUND)
+
+
+def authorize_name(func):
+    """
+    Authorize that a particular certificate has access to any directory
+    containing the repository identified by repo_name.
+
+    :param repo_id: The identifier for the repository
+    :type repo_id: str
+    :rtype: function
+    """
+
+    @wraps(func)
+    def wrapper(repo_id, *args, **kwargs):
+        # will raise an appropriate exception if not found or not authorized
+        name_is_authorized(repo_id)
+        return func(repo_id, *args, **kwargs)
+
+    return wrapper
+
+
+def validate_and_transform_repo_name(username, repo, file_path):
+    """
+    Checks and creates a repo registry id from the username. repo and
+    file path parameters.
+
+    The url path mechanism of flask for url such as
+    <path:username>/<path:repo>/<path:file_path> matches an
+     url such as library/redhat/rhel7.0/test/path to username: library,
+     repo: redhat and path: rhel7.0/test/path. This method removes
+     library and shifts values to username: redhat, repo: rhel7.0
+     and path: test/path.
+
+    :param username: value for <path:username> path component
+    :type  username: basestring
+
+    :param repo: value for <path:rep> path component
+    :type repo: basestring
+
+    :param file_path: value for <path:file_path> path component
+    :type file_path: basestring
+
+    :return: tuple containing transformed/shifted username, repo and file_path if
+             library namespace is contained in username else return the username,
+             repo and file_path without any change
+    :rtype: basestring
+    """
+
+    if 'library' not in username:
```

Comment:
'library' will not be in username unless the user explicitly does docker pull <host>/library/repo. When the user does docker pull <host>/repo, username ends up being repo in this function and then this function returns ('repo', 'manifests', <tag>). For example, when I do this:

```
$ sudo docker pull 127.0.0.1:5001/registry:2
```

This function ends up with username of "registry" and returns ```('registry', 'manifests', '2')``` which will not work.